src/main/java/org/broadinstitute/hellbender/tools/walkers/annotator/StrandArtifact.java

```diff
@@ -0,0 +1,178 @@
+    // give more prior weight to beta (i.e. pseudocount for tails) so that the peak shifts towards 0
+    final int ALPHA = 2;
+    final int BETA = 6;
+
+    // the latent variable z in the strand artifact filter model
+    public enum StrandArtifactZ {
+        ART_FWD, ART_REV, NO_ARTIFACT;
+    }
+
+    @Override
+    public List<String> getKeyNames() {
+        return Arrays.asList(POSTERIOR_PROBABILITIES_KEY, MAP_ALLELE_FRACTIONS_KEY);
+    }
+
+    @Override
+    public void annotate(final ReferenceContext ref,
+                         final VariantContext vc,
+                         final Genotype g,
+                         final GenotypeBuilder gb,
+                         final ReadLikelihoods<Allele> likelihoods) {
+        Utils.nonNull(gb);
+        Utils.nonNull(vc);
+        Utils.nonNull(likelihoods);
+
+        // do not annotate the genotype fields for normal
+        if (g.isHomRef()){
+            return;
+        }
+
+        final double[] pi = new double[]{ 0.025, 0.025, 0.95 };
+
+        // We use the allele with highest LOD score
+        final double[] tumorLods = GATKProtectedVariantContextUtils.getAttributeAsDoubleArray(vc, GATKVCFConstants.TUMOR_LOD_KEY, () -> null, -1);
+        final int indexOfMaxTumorLod = MathUtils.maxElementIndex(tumorLods);
+        final Allele altAlelle =  vc.getAlternateAllele(indexOfMaxTumorLod); // Careful!
+
+        final Collection<ReadLikelihoods<Allele>.BestAllele> tumorBestAlleles = likelihoods.bestAlleles();
+        final int numAltReadsForward = (int) tumorBestAlleles.stream().filter(ba -> !ba.read.isReverseStrand() && ba.isInformative() && ba.allele.equals(altAlelle)).count();
+        final int numAltReadsReverse = (int) tumorBestAlleles.stream().filter(ba -> ba.read.isReverseStrand() && ba.isInformative() && ba.allele.equals(altAlelle)).count();
+        final int numReadsForward = (int) tumorBestAlleles.stream().filter(ba -> !ba.read.isReverseStrand() && ba.isInformative()).count();
+        final int numReadsReverse = (int) tumorBestAlleles.stream().filter(ba -> ba.read.isReverseStrand() && ba.isInformative()).count();
+        final int numReads = numReadsForward + numReadsReverse;
+
+        final double[] unnormalized_posterior_probabilities = new double[StrandArtifactZ.values().length];
+        final double[] maximum_a_posteriori_allele_fraction_estimates = new double[StrandArtifactZ.values().length];
+
+        // Compute the posterior probability of ARTIFACT_FWD and ARTIFACT_REV; it's a double integral over f and epsilon
+        // Note we must calculate the unnormalized probabilities for NO_ART, ARTIFACT_FWD and ARTIFACT_REV in order to get the normalized probabilities
+
+        // the integrand is a polynomial of degree n, where n is the number of reads at the locus
+        // thus to integrate exactly we need (n/2)+1 points
+        final int numIntegPointsForAlleleFraction = numReads / 2 + 1;
```

Comment:
Done. I added a separate step to compute likelihoods for finding MAP estimate for f, which used to piggy back on the double integral. I think the new way is better, as the points that Gauss-Legendre samples may not necessarily cover the points where the posterior probability is high.